[RFC] on new lifecycle methods

paulmelnikow opened the pull request:
Thought I would open a proposal for this in the style of #1429 / #1413.

This aims to improve replace Nock's lifecycle methods with new ones which better fit real-life use cases from my experience with the library. It gives them clearer, less ambiguous names. I include a solution to #884 which I've reopened.

These use cases are the ones that make sense to me, though I'm curious if other folks have different use cases which should be considered as "first class." Though I'm not really _removing_ anything in the API, so if folks have an unusual use case where they need to do things a different way, that's still possible.

I'd like to ship these in 11.x, though we could consider backporting some of them to 10.x (minus the deprecation warnings).

These encourage changes to the installed test code base, which is something I don't take lightly. Before we release it, I'd like to validate the new interface with beta users, and be really sure that it makes their tests clearer and seems worth their effort to update.

gr2m commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,65 @@
+4. Add `nock.assertAll()` which does the equivalent of
+   `scopes.forEach(scope => scope.done())`. This probably should be invoked
+   in the test itself.
+5. Rename `scope.done()` to `scope.assert()` to harmonize with `assertAll()`
```
I’m not sure about the naming of `scope.assert()`, it reads as if I should pass in something. Maybe `scope.assertDone()` and `scope.assertAllDone()`?

paulmelnikow replied:
Hmm yea it does seem like it wants an argument. `nock.assertAllDone()` and `scope.assertDone()` could work. It's a little long but maybe that's fine.

For someone encountering the test code for the first time, it's definitely clearer for than `done()`.

gr2m replied:
I’m a fan of clear method names, even if they are long. Code readability is king 👑

paulmelnikow replied:
Are there any other mock libraries with interfaces you like? I'm curious to compare. I looked at Sinon, but it doesn't really have assertion methods like this.

"Done" isn't the clearest either. Really what we want to say is "expectations fulfilled" or "requests made" or "mocks consumed." I'm 100% agreed that clear long names are better than short names, though I don't want the names to be longer than necessary.

paulmelnikow replied:
Another option is `nock.verifyAll()` / `scope.verify()`.

I don’t think it has the same problem as `assert` of wanting an argument.

Or: `nock.verifyAllRequests()` / `scope.verifyRequest()`

paulmelnikow replied:
I'll keep thinking about it too.

gr2m replied:
how about expressing that all interceptors have been used up, the naming would align with the READ THIS! section in the readme: https://github.com/nock/nock#read-this---about-interceptors

sth like `nock.allInterceptorsUsed()`? 

I wonder if that method is necessary at all, maybe we should reduce the amount of APIs. 

Maybe something like the `.remainingInterceptors()` method would be enough, then I do the assertion myself in my tests?

paulmelnikow replied:
`assertAllInterceptorsUsed()` seems good. Though is it scopes or interceptors we're talking about? Maybe `assertAllMocksUsed()` is better.

We do have `nock.pendingMocks()`, so it's possible if people want control of the assertion they can add it.

Though I would rather not write my own assertion here. It ends up being boilerplate code that has to be replicated across each library. Nock is in a good position to write helpful error messages. Down the line you could imagine saying something more detailed like, "GET http://example.test:80/ expected two calls but only got one", which a caller wouldn't be able to do. I don't like adding unnecessary APIs, though I think nock is the place where this assert function should be implemented.

paulmelnikow replied:
I changed `nock.assertAll()` to `nock.assertAllMocksUsed()` and `scope.assert()` to `scope.assertMocksUsed()`.

gr2m replied:
now seeing that, I wonder why one of the method is called `assertAllMocksUsed` and the other `assertMocksUsed`. Wouldn’t it be more consistent to call both `assertMocksUsed` and the "scope" of the method call would be clear based on wether you call it on a scope instance or the global `nock` object?

paulmelnikow replied:
It’s a good question. If the names are the same I think it’s too likely someone will call the wrong one. Since `scope.done()` was the way to do this for a long time, I don’t want to provide two different ways which are as confusable as `scope.assertMocksUsed()` and `nock.assertMocksUsed()`. Especially if those two functions are doing different things.

paulmelnikow replied:
Another option: `scope.assertUsed()`, `nock.assertAllMocksUsed()`.

gr2m replied:
> Especially if those two functions are doing different things

Wouldn’t both functions do the same thing? `nock.assertMocksUsed()` would check globally, `scope.assertMocksUsed()` would check for only that scope?

paulmelnikow commented:
For the reset case, I wrote:

`nock.restore(); nock.cleanAll(); nock.enableNetConnect()`

To get back to the initial state, it also currently needs:

`nock.activate()`

paulmelnikow commented:
#1465 adds an `afterEach` hook to Nock's test, using those four methods. So let's validate that those are the right ones 😁

paulmelnikow commented:
I rewrote most of the text of this proposal, updated method names based on [this discussion](https://github.com/nock/nock/pull/1441#discussion_r253346600), and proposed a consolidated API for controlling nock's handling of unmocked requests.

It's clear there is a lot of stuff to cover here!  👀 

To make this easier to digest, I've split it up into four sub-proposals, each dealing with one of the main use cases:

1. Resetting
2. Temporary deactivation
3. Assertions
4. Handling unmocked requests

So far we've talked the most about (3) so maybe we could try to approve that one first?

gr2m commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,186 @@
+
+| Use case                                                                | Code                                                                        | Assessment                                                                                                                     |
+| ----------------------------------------------------------------------- | --------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
+| Assert that all mocks have been satisfied                               | `scopes.forEach(scope => scope.done())`                                     | Meh. This requires keeping track of all the scopes.                                                                            |
```
Could you maybe remove the `Meh. ` from the assessments? Maybe it’s just me, but somehow it has a negative vibe and I wouldn’t like it to set a tone for communication for the Nock community

paulmelnikow replied:
What I meant was, this interface is passable; it's not confusing but it's not ideal, either. I'll say that instead.

paulmelnikow replied:
On reflection, thanks for the nudge to keep the language sounding positive. 👍

gr2m commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,186 @@
+#### Discussion
+
+`nock.simulateUnreachability()` emits an unreachable-like error through normal
+HTTP error channels.
```
maybe `.simulateConnectionError()`?

leggsimon replied:
I wonder whether this should be available on a scope? So for example

```js
nock('https://google.com')
	.get('/')
	.simulateConnectionError();
```

Which would allow for some more granular control over which requests produce a network error and which don’t

paulmelnikow replied:
That's an interesting idea, though there would also need to be a way to configure unreachability for unmocked requests.

I think `.replyWithError()` could be used for this purpose today, and would want to think about whether it's worth adding a second way to do it.

leggsimon replied:
Yeah, I think it’s totally reasonable to also want to turn off all network connections as well. But yeah, I think it’s nice to be able to test how your application responds to one API being down/unavailable without it being blanketly based on the host.

The use case here is that I have two requests that go through an API gateway, in my case `api.ft.com`, behind that are different services which potentially could be individually unavailable. So `api.ft.com/foo` is available but `api.ft.com/bar` uses a different surface under the hood which is down. 

Currently (as far as I know) I could only simulate all of `api.ft.com` being unavailable rather than just `api.ft.com/bar`.

paulmelnikow replied:
Makes sense. Although, doesn't `.replyWithError()` solve that problem?

leggsimon replied:
Yeah, I didn’t actually know that existed until you’d just mentioned it. 

It does, however it might be nice to be able to not have to provide a message and have it just fail in the network and let whatever library (eg node-fetch) build the correct network error message, that it would if it was real, if that makes sense?

paulmelnikow replied:
Right, makes sense. It's syntactic sugar, though it keeps the dev from needing to know what kind of error is thrown when a server is unreachable.

leggsimon replied:
Yeah, exactly!

gr2m commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,186 @@
+Okay, so I'm not thrilled with these proposals. While these use cases are
+important and need to be distinguished, having similar behavior triggered by
+two different functions is confusing. Let me try to propose an alternative
+with a combined API.
```
Agreed. I feel we can come up with a better way here

gr2m commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,186 @@
+By putting a wildcard match in the `fail` list, you know your tests will
+immediately choke if they try to hit live servers. Since the client library
+and application are kept out of the loop, this _provides confidence that a
+mock is configured for every request made by the application and tests_.
```
`whenUnmocked` feels like it should be an event handler somehow ... 🤔 I’m not sure about this one, I feel there is a better way. Sorry I know that’s not helpful but all I got right now 🤷‍♂️

paulmelnikow commented:
This is awesome feedback. It's going to take me a minute to digest it and respond.

Thanks for sharing the link about Node governance! "Does anyone object" is not how I'm used to working though I agree it makes sense, and could be really good for nock.

In particular ([from that thread](https://twitter.com/jasnell/status/1100082440054304770)):

> - Agreeing with a decision takes generally more commitment and cognitive load than not-objecting to it. It puts significantly less burden on those who may have less interest in the decision than those who have more.
> - What this enables is a far more decentralized, yet still functional, project governance model that allows us to make progress without a project roadmap or project manager.

paulmelnikow commented on RFC-lifecycle.md:
```diff
@@ -0,0 +1,186 @@
+
+Some developers may prefer to call it from an `afterEach()` hook to avoid
+boilerplate. Since not all tests in a suite may use a mock, it should do the
+natural thing when no mocks have been created, which is no-op.
```
Agreed that `assertMocksUsed()` is better than `assertAllMocksUsed()`. 👍 

Right, my concern is offering two identically named methods which act on different scopes. Most developers won't digest the whole API, they'll just absorb the bits they need. If they know the method is `assertMocksUsed()`, they could get unexpected results if they don't realize that calling it on `scope` will have a more limited effect than calling it on `nock`.

Having reflected on this a while, I'm okay with naming them both `assertMocksUsed()`. However I prefer your suggestion of `nock.assertMocksUsed()` for the global one and `scope.assertScopeMocksUsed()` for the limited one:

1. It encourages people to use the global `nock.assertMocksUsed()` where possible.
2. It makes the limited scope of `scope.assertScopeMocksUsed()` 100% clear.
2. It helps clarify that the intent of `scope.assertScopeMocksUsed()` is for cases which can't be handled using the global method.

gr2m replied:
If we do two different method names, then I’d prefer `nock.assertAllMocksUsed` over `nock.assertMocksUsed`. The latter might suggest I can pass an argument

paulmelnikow commented:
Should we check the accepted RFCs into the repository? The artifact of that discussion could be useful to link to from the release notes for users who are upgrading. I propose we do that.

Since this discussion and scope have gotten large, I could pull all the intro material + analysis into one PR (maybe `rfcs/NOCK-001.md`?), and then create additional PRs for each set of proposed changes (`rfcs/NOCK-002.md`, etc).

[Does anyone object?](https://github.com/nock/nock/pull/1441#pullrequestreview-207675544)

gr2m commented:
That’s be great!
